integration-tests/src/test/java/oracle/weblogic/kubernetes/ItHorizontalPodAutoscalerCustomMetrics.java

```diff
@@ -0,0 +1,393 @@
+    clusterResource.getSpec()
+        .serverPod(new ServerPod().resources(
+            new V1ResourceRequirements()
+                .putLimitsItem("cpu", Quantity.fromString("2"))
+                .putLimitsItem("memory", Quantity.fromString("2Gi"))
+                .putRequestsItem("cpu", Quantity.fromString("250m"))
+                .putRequestsItem("memory", Quantity.fromString("768Mi"))));
+    logger.info("Creating cluster {0} in namespace {1}", clusterResName, domainNamespace);
+    createClusterAndVerify(clusterResource);
+
+    // set cluster references
+    domain.getSpec().withCluster(new V1LocalObjectReference().name(clusterResName));
+
+    // verify the domain custom resource is created
+    createDomainAndVerify(domain, domainNamespace);
+
+    // check admin server is up and running for domain1
+    checkPodReadyAndServiceExists(adminServerPodName, domainUid, domainNamespace);
+
+    // check managed servers are up and running for domain1
+    for (int i = 1; i <= replicaCount; i++) {
+      checkPodReadyAndServiceExists(managedServerPrefix + i, domainUid, domainNamespace);
+    }
+  }
+
+
+  /**
+   * Test autoscaling using HPA with custom WLS metrics, collected by Monitoring Exporter
+   * and exposed via Prometheus and Prometheus Adapter.
+   */
+  @DisabledIfEnvironmentVariable(named = "OKD", matches = "true")
+  @Test
+  void testHPAWithMetricsServer() {
+
+    assertDoesNotThrow(() -> installPrometheus(PROMETHEUS_CHART_VERSION,
+        domainNamespace,
+        domainUid), "Failed to install Prometheus");
+    prometheusAdapterHelmParams = assertDoesNotThrow(() -> installAndVerifyPrometheusAdapter(
+        prometheusAdapterReleaseName,
+        monitoringNS, K8S_NODEPORT_HOST, nodeportPrometheus), "Failed to install Prometheus Adapter");
+    // wait till prometheus adapter could get the current custom metrics
+    // total_opened_sessions_exporter_app to make sure it is ready
+    testUntil(withStandardRetryPolicy,
+        () -> verifyCustomMetricsExposed(domainNamespace,"total_opened_sessions_exporter_app"),
+        logger,
+        "Get current total_opened_sessions_exporter_app from prometheus adapter in namespace {0}",
+        domainNamespace);
+
+    int managedServerPort = 8001;
+    Map<String, Integer> clusterNameMsPortMap = new HashMap<>();
+    clusterNameMsPortMap.put(wlClusterName, managedServerPort);
+
+    String ingressClassName = nginxHelmParams.getIngressClassName();
+    List<String> ingressHostList
+        = createIngressForDomainAndVerify(domainUid, domainNamespace, 0, clusterNameMsPortMap,
+        false, ingressClassName, false, 0);
+    // create hpa with custom metrics
+    createHPA();
+    //invoke app 50 times to generate metrics with number of opened sessions > 15
+    verifyMonExpAppAccessThroughNginx(ingressHostList.get(0), 1, nodeportshttp);
+
+    //check hpa scaled up to one more server
+    checkPodReadyAndServiceExists(managedServerPrefix + 3, domainUid, domainNamespace);
```

Comment:
Does the open sessions go down after some time? can we add check for auto scale down?

Reply:
added code for scale down check